storage/merkle/node.go

```diff
@@ -4,22 +4,89 @@ package merkle
 
 import (
 	"github.com/onflow/flow-go/crypto/hash"
+	"golang.org/x/crypto/blake2b"
 )
 
-type node interface{}
+// The node represents a generic vertex in the sparse Merkle trie.
+// CONVENTION:
+//  * This implementation guarantees that within the trie there are
+//    _never_ nil-references to descendants. We completely avoid this by
+//    utilizing `short nodes` that represent any path-segment without branching.
+type node interface {
+	// Hash computes recursively the hash of this respective sub trie.
+	// To simplify enforcing cryptographic security, we introduce the convention
+	// that hashing a nil node is an illegal operation, which panics.
+	Hash() []byte
+}
+
+// nodeTag encodes the type of node when hashing it. Required for cryptographic
+// safety to prevent malleability attacks (replacing a full node with a short node
+// that has identical hash would otherwise be possible).
```

Comment:
In a different thread, @AlexHentschel explained the attack path : a malicious party than knows the key-values set (and therefore the tree) is able to forge a fake but valid proof of non-membership in certain cases, and this by using a collision in the node hash of a full node and a fake short node. 

This assumes :
- a proof is valid with regards to a proof validation algorithm (not implemented yet)
- the tree is expected to support non-membership (exclusion) proofs.

Reply:
thanks Tarak, you are right that this would be a `collision attack`. Updated the wording accordingly:  https://github.com/onflow/flow-go/blob/c781aaf61b5a136c29bbf7546d46a93d82b96843/storage/merkle/node.go#L21-L23